### ISSUE TITLE
Picking a good value for threads-per-threadgroup

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#275]
Different hardware running the same algorithm may need dramatically different threads-per-threadgroup values in order to get good performance. A small device may need a small value so the register file doesn't spill to main memory, but a big device may need a large value to take advantage of all its multiprocessing lanes. However, an author doesn't know which value is a good value to pick.

This problem is not specific to WebGPU, but we do have a goal of portable performance.

It's kind of doubly bad because the current design of WebGPU bakes in these values as literals into the shader. Therefore, even if the application could divine good values at runtime, it would have to rewrite its shader to use them.

Similarly, each implementation/device has its own thresholds of acceptable values, and these thresholds are specific to each shader. Therefore, compiling a compute shader on one platform may succeed, but on another platform may fail because it ran up against an implementation-specific limit.

**Comment by litherum** [comment]
One thing that might help, though not completely solve the problem, is if the implementation can suggest values that might work well to the author. For example, [threadExecutionWidth](https://developer.apple.com/documentation/metal/mtlcomputepipelinestate/1414911-threadexecutionwidth?language=objc) in Metal, and the [cudaOccupancyMaxPotentialBlockSize](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__HIGHLEVEL.html#group__CUDART__HIGHLEVEL_1gee5334618ed4bb0871e4559a77643fc1) family of functions.

**Comment by litherum** [comment]
Or if the values baked into the shader are optional, and unspecified instructs the runtime to pick good values automatically? If its automatic, there needs to be some reflection information so the implementation knows how their barriers will work

**Comment by Kangz** [comment]
Agreed that this is a pain point. Showing the execution width on the DeviceDescriptor would help, and we can make sure that things run portably (but maybe a bit slower than they could on some hardware) by having a limits for the amount of shared memory used etc.

**Comment by kainino0x** [comment]
I don't think there is any way to make this automatic. The optimal values are not just hardware dependent, they are also highly algorithm dependent. I think the best we can do here without a huge innovation is to expose the hardware parameters just like the underlying APIs do.

**Comment by kvark** [comment]
@kainino0x Is any other API (but Metal) expose the wavefront size? Metal is also different in a way that the user sets up the work group size from the API side in `dispatch`, while in Vulkan and D3D12 that comes from the shader itself. So Metal is consistent with itself, and others are consistent, but if we borrow just the hint semantics from Metal without the API setup part, we'll make a mess.

I'm concerned about what the user would need to do:
  1. write the shader off-line based on an unknown group size
  2. at run-time query the optimal group size
  3. patch the shader with this optimal group size (?!)
  4. create a pipeline from the shader

The step (3) doesn't seem quite nice. @Kangz mentioned that having specialization constants would address this. So looking forward to see how this is going to work in practice.

**Comment by kainino0x** [comment]
Good question. I'm not aware of apis in Vulkan/D3D12 which expose it. We could need our own table to look up in.

That said, I'm not sure the wavefront size is necessarily useful by itself or could end up giving a false signal. For really good tuning, AFAIK (not an expert), it's usually necessary to optimize manually against many configurations (e.g. each major architectural revision from each vendor).

**Comment by RafaelCintron** [comment]
[Wave Intrinsics](https://github.com/Microsoft/DirectXShaderCompiler/wiki/Wave-Intrinsics) are available in Shader Model 6.

**Comment by dneto0** [comment]
>  I'm not aware of apis in Vulkan/D3D12 which expose it.

Vulkan 1.1 added [VkPhysicalDeviceSubgroupProperties](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#VkPhysicalDeviceSubgroupProperties)  which has a subgroupSize member.
That has to match the SubgroupSize execution mode literal value specified in the SPIR-V shader (SPIR-V 1.1 or later).

The Vulkan required values are very much lowest-common demoniator:
fragment shaders, the value is at least 4.   For compute shaders it's at least 1 (time to party!).

I agree that generally tuning subgroup size will get you a lot but is very hardware and algorithm dependent.

**Comment by dneto0** [comment]
Wait a second, I think there's some confusion here.

There are two concepts at play:

- OpenCL/Vulkan workgroup =~  Metal threadgroup 
- OpenCL/Vulkan subgroup =~ Metal SIMD group.  The size of this is returned by threadExecutionWidth. See [this help](https://developer.apple.com/documentation/metal/about_threads_and_threadgroups?language=objc)

Both kinds of execution scopes can vary in size.  The values they can have, most especially the workgroup size can vary according to the shader being compiled.  That's why the Cuda query depends on the kernel function.  Similarly OpenCL provides APIs that depend on the compiled kernel object.
See [5.9 Kernel Objects](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_API.html#_kernel_objects) particularly clGetKernelWorkGroupInfo (and table 30 of allowed queries) and clGetKernelSubGroupInfo (and table 31 for allowed queries)
